Implementation of Object.getOwnPropertyDescriptors.

jordonwii opened the pull request:
Fixes #254 

This is a rough implementation based on a combination of getOwnPropertyDescriptor as a reference, and what I could intuit from a day of digging. 

I'd appreciate as much feedback as possible on how I did this. I used existing functions I was aware of when possible, but if there are better choices of helper functions here, please do let me know. I also did the tests all in javascript (similar to what I could find on getOwnPropertyDescriptor), and tried to make them as thorough as possible.

msftclas commented:
Hi __@jordonwii__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Jordon Wing). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

ianwjhalliday commented on test/Object/getOwnPropertyDescriptors.js:
```diff
@@ -0,0 +1,99 @@
+        f();
+    }
+    catch(e) {
+        WScript.Echo("PASS");
```
It would be better to verify what type of error is thrown, since the spec specifies this.  E.g. TypeError or ReferenceError.

Further, I always recommend that the error string is also compared, to verify that we are getting the expected error rather than some other error of the same type (TypeError is very common).

This is made easier with the UnitTestFramework's assert.throws() method.  Look for other examples of usage of UTF and assert.throws().

ianwjhalliday commented on test/Object/getOwnPropertyDescriptors.js:
```diff
@@ -0,0 +1,99 @@
+]
+
+for (var i in tests) {
+    CatchAndWriteExceptions(tests[i]);
```
This sort of run all tests regardless of exceptions behavior is also made easy by using UnitTestFramework

jordonwii commented:
Apologies for the broken push - I didn't expect that branch to show up here before I explicitly did something somewhere to update the PR. 

In any case, I rewrote my existing tests in the UTF (which is *much* better). The only spec I could find was the one linked to in the issue, and that had one test which I included a slightly modified copy of in my tests.

Also, from the explanation of the property terminology in #254, I can definitely improve that code. It looks like propertyRecord is unused, so I can remove that, and the "propName" variable should be called "propKey".

tcare commented on test/Object/rlexe.xml:
```diff
@@ -195,6 +195,11 @@
   </test>
   <test>
     <default>
+      <files>getOwnPropertyDescriptors.js</files>
```
Make sure to pass "-args summary -endargs" as the flags. That way we can do away with the baseline altogether.

ianwjhalliday commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -817,6 +817,64 @@ namespace Js
+            {
+                return result;
+            }
+        }
```
Not sure if all our built-ins need to do this.  Feels like either they all would or not though.  @Yongqu is this necessary and correct?

jordonwii replied:
FWIW, that code is a copy of the code used by `EntryGetOwnPropertyDescriptor` (the only difference is what operation is passed to `InvokeBuiltInOperationRemotely`).

ianwjhalliday commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -817,6 +817,64 @@ namespace Js
+            newPropertyDescriptor.SetWritable(true);
+            newPropertyDescriptor.SetValue(JavascriptObject::GetOwnPropertyDescriptorHelper(obj, propName, scriptContext));
+
+            JavascriptOperators::DefineOwnPropertyDescriptor(resultObj, propertyRecord->GetPropertyId(), newPropertyDescriptor, true, scriptContext);
```
I believe you can use `SetProperty` instead, to avoid creating a descriptor, since the attributes are all true (their defaults).  Not sure off the top of my head if you should use `resultObj->SetProperty(...)` or `JavascriptOperators::SetProperty(...)` though.  Take a look at what the JavascriptOperators API does to see if it is necessary.  I think it isn't.

jordonwii replied:
It didn't seem necessary to me, so I did use `resultObj->SetProperty(...)`.

ianwjhalliday replied:
Great!

ianwjhalliday commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -817,6 +817,64 @@ namespace Js
+        Var ownPropertyKeys = JavascriptOperators::GetOwnPropertyKeys(obj, scriptContext);
+        JavascriptArray *ownPropsArray = JavascriptArray::FromVar(ownPropertyKeys);
+
+        RecyclableObject* resultObj = scriptContext->GetLibrary()->CreateObject(true);
```
Since we know how many keys there are up front, we could optimize this object by specifying how many properties it will have during creation.  This would then add slots for those properties inline in the allocation for the object (instead of requiring an auxiliary slot array as a second allocation).

We can probably also give the type up front too but I'm not clear on whether we can always share the type from the source object.  E.g. the source object might have special internal properties on its type which shouldn't be on the result object.

Something to make a note of but probably not a priority here.

jordonwii replied:
I passed in the length of `ownPropsArray` as the `requestedSlots` parameter - is that the correct way to make that optimization? If we do this, what happens if the user then tries to define a property on resultObj after we return it?

I'm guessing the way to specify the type would be the `prototype` parameter to CreateObject?  What would the type be, here?

ianwjhalliday replied:
`requestedSlots` is correct.  This means we will allocate the object with its header and its initial slots in one allocation.  If later more properties are added to it they will move to an auxiliary slot array.  Hmm, that could be bad if there are a lot of properties.  Let me think about this more.  It will be fine to take this as is and change it later if this is a problem.

The type is a lot more complicated.  A type is the mapping of property names to property slot indexes in the object's slot array.  Every object has a pointer to its type in its header data.  However, it is non-trivial to know if a type can be shared.  Deep knowledge of the type system would be best to decide how to handle this.  I'll ask around.  I think it would be fine to not bother with the type, but if it is safe and cheap to do it would be best to share.

jordonwii commented:
I incorporated all of the comments left on the previous diff (except the remote dispatch one, since that's unresolved). 

Is the ARM failure normal, or is it because of my change? The only error message in the Jenkins log I can find isn't clear.

ianwjhalliday commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -817,6 +817,60 @@ namespace Js
+            JavascriptConversion::ToPropertyKey(propKey, scriptContext, &propertyRecord);
+
+			Var newDescriptor = JavascriptObject::GetOwnPropertyDescriptorHelper(obj, propKey, scriptContext);
+			resultObj->SetProperty(propertyRecord->GetPropertyId(), newDescriptor, PropertyOperation_None, nullptr);
```
Just noticed there are tabs here.  Please convert tabs to spaces.  Sorry, a nitpick convention we maintain.

jordonwii replied:
Woops. Didn't realize those made it in.

ianwjhalliday commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -817,6 +817,60 @@ namespace Js
+        RecyclableObject* resultObj = scriptContext->GetLibrary()->CreateObject(true, (Js::PropertyIndex) ownPropsArray->GetLength());
+        
+        PropertyDescriptor propDesc;
+        Var propKey;
```
The arm failures appear to be due to extra static analysis.  There is a warning, and warnings are errors in our build.  The compiler believes it is possible for propKey to be used while uninitialized.

It might be the case that DirectGetItemAt doesn't always set propKey to a value.  Should be in this case, but the compiler probably can't see that.

Initializing propKey to nullptr here will probably silence the warning.

I don't know why this is only showing up on the ARM test and release builds.

jordonwii replied:
Makes sense - is there a way for me to verify that the build is working locally? I don't seem to be able to build the ARM target.

ianwjhalliday replied:
You need a particular SDK installed to get the ARM toolset.  I'm not familiar with which SDK though.  Looks like it builds now, so I wouldn't worry about it.

ianwjhalliday commented on test/Object/getOwnPropertyDescriptors.js:
```diff
@@ -0,0 +1,89 @@
+        name: "Object has getOwnPropertyDescriptors method",
+            body: function() {
+                assert.isTrue(Object.hasOwnProperty("getOwnPropertyDescriptors"), 'Object.hasOwnProperty("getOwnPropertyDescriptors")');
+
```
Need more tests here for the spec details of the property and function object, i.e. the property attributes are set correctly and the length property of the function is correct.  There are examples of such tests in some of the other es6 tests.

ianwjhalliday commented on test/Object/getOwnPropertyDescriptors.js:
```diff
@@ -0,0 +1,89 @@
+        body: function() {
+            assert.throws(function() {
+                WScript.Echo(Object.getOwnProperties());
+            }, TypeError, "");
```
Please fill in the error message for the assertion to verify that we are seeing the expected error rather than some other TypeError.

ianwjhalliday commented on test/Object/getOwnPropertyDescriptors.js:
```diff
@@ -0,0 +1,89 @@
+                    enumerable: false,
+                    value: "fooAllFalse",
+                    writable: false
+                }
```
Should also have some properties with symbol names and with integer names, since these property names are handled differently than string names.

ianwjhalliday commented:
Implementation looks great.  I'd like to see more tests before we merge it.  I apologize for not having thoroughly looked through the tests until now.

jordonwii commented:
Thanks for the comments. Assuming the ARM build now works, I think I addressed all of them in the latest commit. The tests now cover properties of the method as well as Symbol and Number names for properties.

ianwjhalliday commented:
@jordonwii quick update, this change is ready to go imo.  I just kicked off internal tests.  Last thing I want to consider before merging is whether this should be on by default or should be guarded behind a flag.

I don't see any functional reason to guard it, but there may be other reasons to do so.  Once I have an answer for that I will merge.

ljharb commented:
@ianwjhalliday it's stage 4, so the only reason to put it behind a flag is if there's serious web compatibility concern. MS TC39 reps stated there was no such concern with this method, so I hope it won't end up behind a flag.

dilijev commented on test/Object/getOwnPropertyDescriptors.js:
```diff
@@ -1,89 +1,166 @@
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+
```
nit: blank line

dilijev commented on test/Object/getOwnPropertyDescriptors.js:
```diff
@@ -1,89 +1,166 @@
+]
+
+
 testRunner.runTests(tests, { verbose: WScript.Arguments[0] != "summary" });
```
nit: newline at end of file

dilijev commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -855,7 +855,7 @@ namespace Js
         
         PropertyDescriptor propDesc;
-        Var propKey;
+        Var propKey = nullptr;
```
Thanks. I like to see this. :)

dilijev commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -817,6 +817,60 @@ namespace Js
+
+        RecyclableObject* obj = nullptr;
+
+
```
nit: remove extra blank line

dilijev commented on test/Object/rlexe.xml:
```diff
@@ -195,6 +195,13 @@
     <default>
+      <files>getOwnPropertyDescriptors.js</files>
+        <compile-flags>-args summary -endargs</compile-flags>
+        <tags></tags>
```
You can omit the empty <tags>...</tags> line as it has the reasonable default of being empty. :)

dilijev commented:
@jordonwii Thanks for this. Before merging, we prefer history to be compressed via `git rebase -i` or `git merge --squash` to a single commit, unless there is future value in having multiple commits in a pull request. (Especially when some of those commits failed tests.) We like to keep the history as clean as possible.

jordonwii commented:
I can do that. Would that cause the previous comments on specific diff lines to lose context, though, or does GH preserve the diffs if I rewrite history?

dilijev commented:
@jordonwii It will fold the comments and say they are on an "outdated diff" so you'll lose context on any open discussions IF the lines changed, but hopefully that means you resolved the discussions anyway.

dilijev commented on test/Object/getOwnPropertyDescriptors.js:
```diff
@@ -0,0 +1,159 @@
+    {
+        name: "The resulting get and set are identical with the original get and set.",
+        body: function() {
+            // This test is taken from https://github.com/tc39/proposal-object-getownpropertydescriptors/blob/master/test/built-ins/Object/getOwnPropertyDescriptors/has-accessors.js
```
We should probably say "adapted from" since the asserts use our UTF.

dilijev commented:
LGTM!

Yongqu commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -817,6 +817,59 @@ namespace Js
+        }
+
+        Var ownPropertyKeys = JavascriptOperators::GetOwnPropertyKeys(obj, scriptContext);
+        JavascriptArray *ownPropsArray = JavascriptArray::FromVar(ownPropertyKeys);
```
Can you add a JavascriptArray::Is(ownPropertyKeys) check? I am a bit afraid of potential change here in the future.

ianwjhalliday replied:
Interesting.  If `GetOwnPropertyKeys` is always returning a `JavascriptArray*` we should ideally change the return type to enforce strong typing and break the callsites if the return type changes in the future.

No reason to make this a runtime check if it can be a compile time enforcement.

Yongqu replied:
reflect.ownKeys (26.1.10) has return CreateArrayFromList(keys) which creates a builtin normal array, so from spec perspective it should be a valid normal JavascriptArray. However we don't know if this could change in the future an assert might not be able to catch all the scenario, especially people run retail on the web. I think we should have a runtime check. 

---
In reply to: [69818695](https://github.com/Microsoft/ChakraCore/pull/1202#discussion_r69818695) [](ancestors = 69818695)

ianwjhalliday replied:
Sure but this is controlled by our runtime and so if currently our runtime always returns a JavascriptArray*, we can simply be more specific in the return type, to avoid being confused about whether the result is always an array or not at the call sites.

The FromVar conversion would then be unnecessary.

Also I just remembered that FromVar performs the assert check already, so the assert added here is actually redundant as it stands.

tcare replied:
Unless we change the contract to be JavascriptArray*, a Var from JavascriptOperators should have an ::Is check. Most JavascriptOperator methods return Var though.

Usually I would do this for anything stronger than a RecyclableObject* assumption. It's quite important for anything we don't end up catching from asserts, especially if it would require a different error path.

Yongqu commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -817,6 +817,59 @@ namespace Js
+
+        for (uint i = 0; i < ownPropsArray->GetLength(); i++)
+        {
+            ownPropsArray->DirectGetItemAt(i, &propKey);
```
check return value.

jordonwii commented:
@Yongqu I added Asserts to handle the failure cases in the implementation. Should any of those *not* be asserts?

dilijev commented:
@jordonwii You just need to ask yourself for each case, since Asserts are not present in Release builds:

* Should we crash the process (unrecoverable state results from this code path)? Then we should FailFast instead of Assert
* Was this simply an unexpected code path that we want to capture in debug builds: Then we should Assert
  * Remember that the Assert will not be present in a Release build, so the rest of the code should handle that case in some reasonable way.

dilijev commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -858,10 +859,12 @@ namespace Js
         {
-            ownPropsArray->DirectGetItemAt(i, &propKey);
+            BOOL getPropResult = ownPropsArray->DirectGetItemAt(i, &propKey);
+            Assert(getPropResult);
```
How should we handle this case in Release builds (when the Assert is compiled out)? Is `getPropResult == false` something that should **never** happen if the code is written correctly?

ianwjhalliday replied:
Good question.  I was assuming that `JavascriptOperators::GetOwnPropertyKeys` always returns a simple JavascriptArray, created by our runtime, so we can make assumptions about it.  However I have not verified this.  @jordonwii is that what you found?

Also I do not know about the assert below for `propertyKeyResult`.  Does `JavascriptConversion::ToPropertyKey` always succeed with a result from `JavascriptOperators::GetOwnPropertyKeys`?

jordonwii replied:
Assuming the object isn't a `Proxy`, `JavascriptOperators::GetOwnPropertyKeys` just returns the result of `JavascriptObject::CreateKeysHelper`, which is always a `JavascriptArray`.

I understood @Yongqu to be concerned about catching a change in something along the chain of calls after `JavascriptOperators::GetOwnPropertyKeys` that such that it *doesn't* ultimately return an Array (hence an assert being appropriate here). I just noticed that `JavascriptArray::FromVar` would already catch such a case, though, since `FromVar` includes the assert:
`AssertMsg(Is(aValue), "Ensure var is actually a 'JavascriptArray'");`

So the one I added is redundant, it looks like. 

Both `JavascriptConversion::ToPropertyKey` and `ownPropsArray->DirectGetItemAt` return a BOOL indicating success/failure, although AFAICT the implementation of `JavascriptConversion::ToPropertyKey` *always* succeeds. That doesn't mean it always will, though, and `JavascriptObject::GetValuesOrEntries` currently includes an Assert identical to the one I added (see https://github.com/Microsoft/ChakraCore/blob/master/lib/Runtime/Library/JavascriptObject.cpp#L1202)

ianwjhalliday replied:
This opened up a good discussion in the office :)  @Yongqu @tcare and I got together and discussed the design of the APIs we have.

1. Currently the array you're calling `DirectGetItemAt` is guaranteed by our implementation to have items in order up to its length.  However we cannot convey this guarantee in the type system and so have no compile time guarantee.  In general this may not be true of JavascriptArray, elements can be deleted causing `DirectGetItemAt` to return `FALSE` for example.

 Because we cannot make a compile time guarantee here, we would like to see the return value handled in release builds.  Probably the best thing to do is Assert if it is false, and then skip the loop body and go to the next element in the array, since this shouldn't happen, but we don't want to use `propKey` if it wasn't guaranteed to be set to anything legitimate.

2. `JavascriptConversion::ToPropertyKey` should return `void`.  It is wasting time by having a confusing bool return type when it always succeeds.  This was poor API design but we can address it.  Not in this PR.  I'll open a separate issue.  I took a look and almost all callers of this API ignore the return value.  I think you should do the same and we'll change the return type to void in another change soon.

3. I verified that `JavascriptOperators::GetOwnPropertyKeys` always returns a `JavascriptArray*` regardless of which control path it takes.  `JavascriptOperators` APIs have traditionally been `Var` or `BOOL` returning, high level operations matching closely with the spec.  Returning `Var` makes sense particularly when the JIT leverages these APIs as helper calls.  However `GetOwnPropertyKeys` is used only internally to our runtime (not used by JIT) and all call sites immediately convert its result back to `JavascriptArray*`.  There's no reason we should conver to `Var` and back to `JavascriptArray*`, C++ is strongly typed, we should take advantage of that.

 I will open an issue to track making this improvement to our code base.  Let's leave it out of scope for your PR.  Since we're leaving it out though, your PR should be resilient, so you should have a code path that handles `JavascriptArray::Is(ownPropertyKeys)` returning false.  In that case you could just set ownPropertyKeys to a new zero length array via `scriptContext->GetLibrary()->CreateArray(0)`.

jordonwii replied:
Great. That all makes sense to me. Glad this discussion was useful!

ianwjhalliday commented:
@ljharb indeed it should remain enabled so long as the spec accepts it.  We decided that it should be controlled with a flag so that it can be disabled if a scenario warrants it (e.g. website bug reduction).

@jordonwii can you add a flag to control the inclusion of this API on the Object constructor.  See `ConfigFlagsList.h` and `JavascriptLibrary.cpp` for examples of how we do this for other APIs.  In this case please add a new flag specifically for this one API.  We may consolidate it with other flags in the future but for now a specific flag for this one API is good.

ianwjhalliday commented:
@jordonwii the linux branch merged into master and pending PRs need to be rebased to include commit 606c442.  Could you also rebase to include the latest master please.  You can do this when feedback has settled and you're done making changes and we're ready to merge.

jordonwii commented:
@ianwjhalliday Minor question re: the flag. What I've read about TC39 says that this feature isn't strictly ES{N}, but do we want it disabled when ES6 disabled, anyway, as ES6 and ES7 features are now (i.e. do we want ES6 do be a parent flag of the new one in ConfigFlagsList.h)?

ianwjhalliday commented:
@jordonwii that is an orthogonal issue we can leave alone for now.  That is, go with the status quo, let it be controlled by the ES6 master switch.

Figuring out how to go forward with switch names in the face of the new way TC39 is adding features to the spec is something we haven't done yet.  We have had some discussions but nothing we like has come about yet.

jordonwii commented:
Alright, I added the flag and incorporated the decisions wrt the implementation. I didn't add any automated test for the flag, but verified manually that the flag being false removes getOwnPropertyDescriptors from Object, and passing in -ObjectGetOwnPropertyDescriptors makes all tests pass.

ljharb commented:
@ianwjhalliday @jordonwii this feature is strictly ES2017 / ES8, since it's stage 4 after ES2016 was finalized. Certainly if "ES6" is disabled, then it makes sense to disable everything that comes after that - but I also can't conceive of any scenario where you could realistically disable ES6+ going forward.

tcare commented:
@ljharb ES6 flag control was introduced at a time where compat modes were still a thing in IE. When Edge came along and we no longer had compat modes it just became the default. It just became a catch all for new ES features.

Often 'ES6' in Chakra is misused, including the language team who still calls themselves ES6 :)

As for the flag set itself, we don't have a need to disable an already released standard. The flag set is just used to group things. Individual flags are what we care about, where they will get fed into automation and used for debugging and reducing.

We should really review our usage of 'es6' naming to reflect current standards conventions.

tcare commented on lib/Common/ConfigFlagsList.h:
```diff
@@ -903,6 +903,9 @@ FLAGPR_REGOVR_EXP(Boolean, ES6, ES6RegExSymbols        , "Enable ES6 RegExp symb
 FLAGPR           (Boolean, ES6, ES6Verbose             , "Enable ES6 verbose trace"                                 , DEFAULT_CONFIG_ES6Verbose)
 FLAGPR_REGOVR_EXP(Boolean, ES6, ArrayBufferTransfer    , "Enable ArrayBuffer.transfer"                              , DEFAULT_CONFIG_ArrayBufferTransfer)
+
+FLAGPR           (Boolean, ES6, ObjectGetOwnPropertyDescriptors, "Enable Object.getOwnPropertyDescriptors"              , DEFAULT_CONFIG_ObjectGetOwnPropertyDescriptors)
```
At the very least I would prefix the flag with ES - it's handy to be able to see that this is a new language feature from a distance.

@ianwjhalliday thoughts?

ianwjhalliday replied:
Agree with ES prefix to easily differentiate ES feature flags at a distance.

I think we can just use ES going forward and not include the version since the features won't always clearly make it into a particular ES publication in the new yearly release train model.

tcare commented on test/Object/rlexe.xml:
```diff
@@ -195,6 +195,12 @@
   <test>
     <default>
+      <files>getOwnPropertyDescriptors.js</files>
+      <compile-flags>-args summary -endargs</compile-flags>
```
Passing the flag in here is a good idea too - that way unit tests still pass if you mess with what is on by default.

jordonwii commented:
Problem was that the "dynapogo" tests passed, but the full suite didn't, and I only noticed the results of the former. I fixed that, and added -ObjectGetOwnPropertyDescriptors to rlexe.xml.

jordonwii commented:
The latest push rebased off of master, and includes the Linux merge. If there isn't any more feedback, then I think we just need a verdict on whether to change the name of the flag, and then I can `rebase -i` to squash all my commits together.

dilijev commented:
@jordonwii If you want to avoid the rebase -i interface, you can also use `merge --squash`:

```
git fetch origin
git branch -M issue254_getownpropertydescriptors i254
git checkout -b issue254_getownpropertydescriptors origin/master
git merge --squash i254
git push issue254_getownpropertydescriptors
```

dilijev commented:
Yeah I think that all of those commits probably fit logically into a single commit. Definitely rebase -i or squash IMO

ianwjhalliday commented on lib/Runtime/Library/JavascriptLibrary.cpp:
```diff
@@ -6810,6 +6815,7 @@ namespace Js
 
         REG_OBJECTS_LIB_FUNC(defineProperty, JavascriptObject::EntryDefineProperty);
         REG_OBJECTS_LIB_FUNC(getOwnPropertyDescriptor, JavascriptObject::EntryGetOwnPropertyDescriptor);
+        REG_OBJECTS_LIB_FUNC(getOwnPropertyDescriptors, JavascriptObject::EntryGetOwnPropertyDescriptors);
```
This should be controlled by the flag too

ianwjhalliday commented on lib/Runtime/Library/JavascriptLibrary.cpp:
```diff
@@ -3794,6 +3794,11 @@ namespace Js
             library->AddFunctionToLibraryObject(objectConstructor, PropertyIds::getOwnPropertyDescriptor, &JavascriptObject::EntryInfo::GetOwnPropertyDescriptor, 2));
+        if (scriptContext->GetConfig()->IsObjectGetOwnPropertyDescriptorsEnabled())
+        {
+            scriptContext->SetBuiltInLibraryFunction(JavascriptObject::EntryInfo::GetOwnPropertyDescriptors.GetOriginalEntryPoint(),
```
propertyCount above needs to be incremented.  It is a hint for the number of properties the object is going to have.  I believe we round up to the nearest 4 or 8, but regardless, the number we supply should be exact or at least exactly the maximum possible (with all features enabled)

ianwjhalliday commented:
Few minor things, ES prefix on the flag name and a couple minor details in JavascriptLibrary.cpp

After fixing those I'd say good to squash

jordonwii commented:
@ianwjhalliday Addressed all of your comments and squashed the commits. Looks like everything built successfully. Let me know if anything else needs changing.

dilijev commented:
LGTM

tcare commented:
Looks good!

ianwjhalliday commented:
:+1: :shipit: I'll get this merged.  Some internal test baselines need to change so might take me a moment to get it together and validate it.

ianwjhalliday commented:
Huzzah, it's in! Thank you @jordonwii for all your work here, much appreciated.

tcare commented:
Congrats @jordonwii! Thanks for your contribution.

jordonwii commented:
Awesome. Thank you all for your help, feedback, and explanations, @Yongqu @ianwjhalliday @tcare @dilijev!
